src/thermo/Phase.cpp

```diff
@@ -794,24 +834,50 @@ void Phase::modifySpecies(size_t k, shared_ptr<Species> spec)
             "New species name '{}' does not match existing name '{}'",
                            spec->name, speciesName(k));
     }
-    const shared_ptr<Species>& old = m_species[toLowerCopy(spec->name)];
+    const shared_ptr<Species>& old = m_species[spec->name];
     if (spec->composition != old->composition) {
         throw CanteraError("Phase::modifySpecies",
             "New composition for '{}' does not match existing composition",
             spec->name);
     }
-    m_species[toLowerCopy(spec->name)] = spec;
+    m_species[spec->name] = spec;
     invalidateCache();
 }
 
 shared_ptr<Species> Phase::species(const std::string& name) const
 {
-    return m_species.at(toLowerCopy(name));
+    std::map<std::string, shared_ptr<Species> >::const_iterator it;
+
+    it = m_species.find(name);
+    if (it != m_species.end()) {
+        return it->second;
+    } else if (!m_caseSensitiveSpecies) {
+        std::string nLower = toLowerCopy(name);
+        bool found = false;
+        shared_ptr<Species> sp;
+        for (const auto& k : m_species) {
+            if (toLowerCopy(k.first) == nLower) {
+                if (!found) {
+                    found = true;
+                    sp = k.second;
+                } else {
+                    throw CanteraError("Phase::species",
+                                       "Lowercase species name '{}' is not unique. "
+                                       "Set Phase::caseSensitiveSpecies to true to "
+                                       "enforce case sensitive species names", nLower);
+                }
+            }
+        }
+        return sp;
+    } else {
+        throw CanteraError("Phase::setMassFractionsByName",
+                           "Unknown species '{}'", name);
+    }
```

Comment:
Isn't there a simpler implementation of this method that leans on `speciesIndex` much like `setMoleFractionsByName`? Something like:
```c++
size_t k = speciesIndex(name);
if (k != npos) {
    return m_species.at(speciesName(k));
} else {
    throw CanteraError(...);
}
```
Or is there some other complexity here that I'm missing?

Reply:
Didn't think of that one!